src/System.Private.CoreLib/src/System/Text/Utf8Span.cs

```diff
@@ -0,0 +1,302 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text.Unicode;
+using Internal.Runtime.CompilerServices;
+
+#pragma warning disable 0809  //warning CS0809: Obsolete member 'Utf8Span.Equals(object)' overrides non-obsolete member 'object.Equals(object)'
+
+#pragma warning disable SA1121 // explicitly using type aliases instead of built-in types
+#if BIT64
+using nint = System.Int64;
+using nuint = System.UInt64;
+#else
+using nint = System.Int32;
+using nuint = System.UInt32;
+#endif
+
+namespace System.Text
+{
+    [StructLayout(LayoutKind.Auto)]
+    public readonly ref partial struct Utf8Span
+    {
+        /// <summary>
+        /// Creates a <see cref="Utf8Span"/> from an existing <see cref="Utf8String"/> instance.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Utf8Span(Utf8String? value)
+        {
+            Bytes = Utf8Extensions.AsBytes(value);
+        }
+
+        /// <summary>
+        /// Ctor for internal use only. Caller _must_ validate both invariants hold:
+        /// (a) the buffer represents well-formed UTF-8 data, and
+        /// (b) the buffer is immutable.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private Utf8Span(ReadOnlySpan<byte> rawData)
+        {
+            // In debug builds, we want to ensure that the callers really did validate
+            // the buffer for well-formedness. The entire line below is removed when
+            // compiling release builds.
+
+            Debug.Assert(Utf8Utility.GetIndexOfFirstInvalidUtf8Sequence(rawData, out _) == -1);
+
+            Bytes = rawData;
+        }
+
+        public ReadOnlySpan<byte> Bytes { get; }
+
+        public static Utf8Span Empty => default;
+
+        public bool IsEmpty => Bytes.IsEmpty;
+
+        /// <summary>
+        /// Similar to "== default", but ignores the Length check.
+        /// </summary>
+        private unsafe bool IsNull
```

Comment:
I have looked at this like for 5 minutes and I am not able to figure why checking the Length for zero is not enough in the places where it is used. Do we really need this?

Reply:
It's used internally by the `Split` routines. Imagine that you're trying to split the string `u"foobar,"` by the char `','`. The split will result in two slices: the first slice is `theSpan[0..6]`, and the second slice is `theSpan[7..7]` (since there's an empty string after the comma).

Now imagine that you're trying to split the string `u"foobar"` by the char `','`. The first slice is `theSpan[0..6]`, and the second slice is `nullptr` (since the search term was not found).

The `Split` routines and enumerator rely on detecting the null span specifically (as opposed to any arbitrary zero-length span that contains a non-null reference) so that they know when to halt enumeration. But this is just an implementation detail, which is why the API is internal.

Reply:
The Utf8Span that is Null vs. non-Null should behave exactly the same as far as the public APIs are concerned. Is that right?

This overloading of Utf8Span pointing to null - even when it is done just for internal purpose - looks like a bug farm to me. For example, consider this:

```
using System;
using System.Text;

class Test
{
    static void Work(ReadOnlySpan<byte> span)
    {
        var utf8Span = Utf8Span.UnsafeCreateWithoutValidation(span);

        foreach (var e in utf8Span.Split('a'))
        {        
            Console.WriteLine("Here");
        }
    }

    static void Main()
    {
        Console.WriteLine("default");
        Work(default);

        Console.WriteLine("new byte[0]");
        Work(new byte[0]);
    }
}
```

Actual result:
```
default
new byte[0]
Here
```

Expected result:
```
default
Here
new byte[0]
Here
```

I do not think we should be using null to have a special meaning. Instead, return explicit bool, etc.

Reply:
> Now imagine that you're trying to split the string u"foobar" by the char ','. The first slice is theSpan[0..6], and the second slice is nullptr (since the search term was not found).

Why the split can't return one slice in this case? instead of returning null item?

---
In reply to: [324432306](https://github.com/dotnet/coreclr/pull/26711#discussion_r324432306) [](ancestors = 324432306)

Reply:
> Why the split can't return one slice in this case?

Internally, the `Split` routine needs to figure out the "before" and "after" spans resulting from the split, and it also needs some way to mark the result as "ignore the second slice - there's no useful data". The public API which wraps the internal split logic filters on this and returns the appropriate value to the caller. But as Jan pointed out, it looks like there are some edge cases that fall down in this scenario.

Part of the design of the helper routines is that I'm trying to avoid stack-spilling as much as possible. Will need to reconsider this to see if there's a way to smuggle more flags across without causing enregistration to fall over dead.

Reply:
Fixed as of latest iteration. I've removed the _null_ special-casing and added a test for this specific scenario: see https://github.com/GrabYourPitchforks/corefx/commit/4e47c84a17d12b40f500b8fcf6aff3b0a5a54633.